### ISSUE TITLE
Unable to load my local bam file in pileupjs

### ISSUE THREAD

**Issue by manojbhosale** [hammerlab/pileup.js#405]
I have installed WAMP stack and I am trying to load a BAM file but I am getting below error.

_"q.js:161 Uncaught Monster request: Won't fetch Infinity bytes from ./test-data/molBarcode.bam"_

Am I doing something wrong here?

below is the code snippet from data.js
 
var bamSource = pileup.formats.bam({
  url: './test-data/molBarcode.bam',
  indexUrl: './test-data/molBarcode.bam.bai'
   });

**Comment by armish** [comment]
@manojbhosale @deepaksibar: thanks you both for giving pileup.js a try and reporting this issue to us. We are trying to replicate this issue on our end, but it would be really useful if you guys could let us know what browser you are using and what version of nodejs you have installed on your system.

We used to have trouble with older Safari browsers and those problems were manifesting themselves within q.js calls, just like you both mentioned. If it is not a browser issue, my second guess would be that your BAM files are being served from an HTTP(s) server that doesn't support RANGE requests. 

In any case, as soon as we can track the problem down with your help, we will address this issue.

Thanks!

**Comment by manojbhosale** [comment]
I am using google chrome browser, Apache(v2.2.17), nodejs(v5.4.1) HTTP server.
For file in test data "synth3.normal.17.7500000-7515000.bam" I am successfully able to load it, but for other files with higher coverage ~3000-10000 reads its failing with following error;
"q.js:161 Uncaught Monster request: Won't fetch Infinity bytes from ./test-data/molBarcode.bam"

I am not sure if above HTTP servers support range requests.

Thanks!

**Comment by subji** [comment]
+1

**Comment by armish** [comment]
All right; first of all sorry for the delay in getting back to this. 

I am trying to replicate this issue but having tried a bunch of example BAMs, I still cannot. There is probably something I am missing here; but can anyone share at least a slice of the BAM file with me so that I can try to track down the problem?

In the meantime, it would be really useful if you can verify that your server accepts RANGE requests. Here is a way to do it, FWIW: http://serverfault.com/a/547078

Thanks!
-- Arman

**Comment by manojbhosale** [comment]
Hi Armish,

Thanks for the reply !!

The concerned BAM file is shared at below location:
https://drive.google.com/open?id=0B42X7X6VgvaFQ1JvR3UwMU1wbTQ


Thanks,
Manoj

**Comment by armish** [comment]
Thanks for that file, @manojbhosale — it really helped!

I have a new PR that fixes this issue and is now waiting for code review (#407). We will get this in and make a minor release soon. I will update this issue once that is done.

In the meantime, thanks to your file, I found that the BAI you provided had weird empty contig definitions that were violating some of our assumptions in pileup.js. Re-indexing your BAM file with samtools fixed the issue for me. Just out of curiosity: how did you get those BAI files generated?

**Comment by manojbhosale** [comment]
The BAI file is generated from a one of my custom tool similar to samtools.
Could you please let me know the limitations if any, of pileupjs in terms  of BAM file size it support?
 I am asking this because, when I try to load some BAM files the browser process sometimes take around >1gb memory.
Also please let me know the tentative time line in which i can expect the fix to be part of pileupjs.

Thanks for your help..

-Manoj

**Comment by armish** [comment]
Hi @manojbhosale,

Just did a minor release update (v0.6.4) to get this fix into pileup.js. Thanks for your help identifying and  fixing the problem.

The memory issue might require some more work and hard thinking about other optimizations, but another user found that it was useful to filter BAMs according to the region of interest and then pass it to pileup.js to improve the overall performance and to reduce the load on the browser. See this pull request for more context: #403. Do you think this will help?

Otherwise, let's create a new issue and discuss our options there.
